RAGchain/DB/pickle_db.py

```diff
@@ -54,19 +54,36 @@ def create_or_load(self):
         else:
             self.create()
 
-    def save(self, passages: List[Passage]):
+    def save(self, passages: List[Passage], upsert: bool = False):
         """Saves the given list of Passage objects to the pickle database. It also saves the data to the Linker."""
+
         # save to pickleDB
-        self.db.extend(passages)
+        if os.path.exists(self.save_path):
+            self.load()
+            passages_dict = {passage.id: passage for passage in passages}
+            updated_passages = []
+
+            for existing_passage in self.db:
```

Comment:
check all id in db? Why?

If we want to add 10 passages, but already ingested million passages, we have to check all million passages?
It looks super inefficient.